broadcom_sta fails to build

### Steps To Reproduce
Steps to reproduce the behavior:
1. `nix-build -A linuxKernel.packages.linux_latest_libre.broadcom_sta`

### Build log
```
/build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:1792:20: error: initialization of 'int (*)(struct wiphy *, struct net_device *, int,  u8,  bool,  const u8 *, struct key_params *)' {aka 'int (*)(struct wiphy *, struct net_device *, int,  unsigned char,  _Bool,  const unsigned char *, struct key_params *)'} from incompatible pointer type 's32 (*)(struct wiphy *, struct net_device *, u8,  bool,  const u8 *, struct key_params *)' {aka 'int (*)(struct wiphy *, struct net_device *, unsigned char,  _Bool,  const unsigned char *, struct key_params *)'} [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wincompatible-pointer-types-Werror=incompatible-pointer-types8;;]
 1792 |         .add_key = wl_cfg80211_add_key,
      |                    ^~~~~~~~~~~~~~~~~~~
/build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:1792:20: note: (near initialization for 'wl_cfg80211_ops.add_key')
/build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:1793:20: error: initialization of 'int (*)(struct wiphy *, struct net_device *, int,  u8,  bool,  const u8 *)' {aka 'int (*)(struct wiphy *, struct net_device *, int,  unsigned char,  _Bool,  const unsigned char *)'} from incompatible pointer type 's32 (*)(struct wiphy *, struct net_device *, u8,  bool,  const u8 *)' {aka 'int (*)(struct wiphy *, struct net_device *, unsigned char,  _Bool,  const unsigned char *)'} [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wincompatible-pointer-types-Werror=incompatible-pointer-types8;;]
 1793 |         .del_key = wl_cfg80211_del_key,
      |                    ^~~~~~~~~~~~~~~~~~~
/build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:1793:20: note: (near initialization for 'wl_cfg80211_ops.del_key')
/build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:1794:20: error: initialization of 'int (*)(struct wiphy *, struct net_device *, int,  u8,  bool,  const u8 *, void *, void (*)(void *, struct key_params *))' {aka 'int (*)(struct wiphy *, struct net_device *, int,  unsigned char,  _Bool,  const unsigned char *, void *, void (*)(void *, struct key_params *))'} from incompatible pointer type 's32 (*)(struct wiphy *, struct net_device *, u8,  bool,  const u8 *, void *, void (*)(void *, struct key_params *))' {aka 'int (*)(struct wiphy *, struct net_device *, unsigned char,  _Bool,  const unsigned char *, void *, void (*)(void *, struct key_params *))'} [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wincompatible-pointer-types-Werror=incompatible-pointer-types8;;]
 1794 |         .get_key = wl_cfg80211_get_key,
      |                    ^~~~~~~~~~~~~~~~~~~
/build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:1794:20: note: (near initialization for 'wl_cfg80211_ops.get_key')
/build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:1795:28: error: initialization of 'int (*)(struct wiphy *, struct net_device *, int,  u8,  bool,  bool)' {aka 'int (*)(struct wiphy *, struct net_device *, int,  unsigned char,  _Bool,  _Bool)'} from incompatible pointer type 's32 (*)(struct wiphy *, struct net_device *, u8,  bool,  bool)' {aka 'int (*)(struct wiphy *, struct net_device *, unsigned char,  _Bool,  _Bool)'} [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wincompatible-pointer-types-Werror=incompatible-pointer-types8;;]
 1795 |         .set_default_key = wl_cfg80211_config_default_key,
      |                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:1795:28: note: (near initialization for 'wl_cfg80211_ops.set_default_key')
cc1: some warnings being treated as errors
make[2]: *** [/nix/store/vc4xd7fz34rvr5i64j2d0w4gxhgmyqbx-linux-6.1-dev/lib/modules/6.1.0-gnu/source/scripts/Makefile.build:250: /build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.o] Error 1
make[1]: *** [/nix/store/vc4xd7fz34rvr5i64j2d0w4gxhgmyqbx-linux-6.1-dev/lib/modules/6.1.0-gnu/source/Makefile:1992: /build/broadcom-sta] Error 2
make[1]: Leaving directory '/nix/store/vc4xd7fz34rvr5i64j2d0w4gxhgmyqbx-linux-6.1-dev/lib/modules/6.1.0-gnu/build'
make: *** [Makefile:161: all] Error 2
error: builder for '/nix/store/pi4n8q67lpw4qgirm767cq56z3rk4xbc-broadcom-sta-6.30.223.271-6.1.drv' failed with exit code 2;
       last 10 log lines:
       > /build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:1794:20: note: (near initialization for 'wl_cfg80211_ops.get_key')
       > /build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:1795:28: error: initialization of 'int (*)(struct wiphy *, struct net_device *, int,  u8,  bool,  bool)' {aka 'int (*)(struct wiphy *, struct net_device *, int,  unsigned char,  _Bool,  _Bool)'} from incompatible pointer type 's32 (*)(struct wiphy *, struct net_device *, u8,  bool,  bool)' {aka 'int (*)(struct wiphy *, struct net_device *, unsigned char,  _Bool,  _Bool)'} [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wincompatible-pointer-types-Werror=incompatible-pointer-types8;;]
       >  1795 |         .set_default_key = wl_cfg80211_config_default_key,
       >       |                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       > /build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.c:1795:28: note: (near initialization for 'wl_cfg80211_ops.set_default_key')
       > cc1: some warnings being treated as errors
       > make[2]: *** [/nix/store/vc4xd7fz34rvr5i64j2d0w4gxhgmyqbx-linux-6.1-dev/lib/modules/6.1.0-gnu/source/scripts/Makefile.build:250: /build/broadcom-sta/src/wl/sys/wl_cfg80211_hybrid.o] Error 1
       > make[1]: *** [/nix/store/vc4xd7fz34rvr5i64j2d0w4gxhgmyqbx-linux-6.1-dev/lib/modules/6.1.0-gnu/source/Makefile:1992: /build/broadcom-sta] Error 2
       > make[1]: Leaving directory '/nix/store/vc4xd7fz34rvr5i64j2d0w4gxhgmyqbx-linux-6.1-dev/lib/modules/6.1.0-gnu/build'
       > make: *** [Makefile:161: all] Error 2
       For full logs, run 'nix log /nix/store/pi4n8q67lpw4qgirm767cq56z3rk4xbc-broadcom-sta-6.30.223.271-6.1.drv'.
```

### Additional context

### Notify maintainers
<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

@X9VoiD maybe

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.81, NixOS, 23.05 (Stoat), 23.05.20221205.6e51c97`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.1`
 - channels(root): `"nixos-21.11pre299825.d8079260a30"`
 - nixpkgs: `/run/current-system/nixpkgs`

--- a/pkgs/os-specific/linux/broadcom-sta/default.nix
+++ b/pkgs/os-specific/linux/broadcom-sta/default.nix
@@ -45,6 +45,8 @@ stdenv.mkDerivation {
     ./linux-5.18.patch
     # source: https://gist.github.com/joanbm/207210d74637870c01ef5a3c262a597d
     ./linux-6.0.patch
+    # source: https://gist.github.com/joanbm/94323ea99eff1e1d1c51241b5b651549
+    ./linux-6.1.patch
     ./pedantic-fix.patch
     ./null-pointer-fix.patch
     ./gcc.patch

--- a/pkgs/os-specific/linux/broadcom-sta/linux-6.1.patch
+++ b/pkgs/os-specific/linux/broadcom-sta/linux-6.1.patch
@@ -1,0 +1,83 @@
+From a63a5f70e5cf05f6bce4cda2e0dd67462e1d76a5 Mon Sep 17 00:00:00 2001
+From: Joan Bruguera <joanbrugueram@gmail.com>
+Date: Mon, 29 Aug 2022 00:06:53 +0200
+Subject: [PATCH] Tentative patch for broadcom-wl 6.30.223.271 driver for Linux 6.1-rc1
+
+Applies on top of all the patches applied to broadcom-wl-dkms 6.30.223.271-35 on Arch Linux
+---
+ src/wl/sys/wl_cfg80211_hybrid.c | 21 +++++++++++++--------
+ 1 file changed, 13 insertions(+), 8 deletions(-)
+
+diff --git a/src/wl/sys/wl_cfg80211_hybrid.c b/src/wl/sys/wl_cfg80211_hybrid.c
+index 4fef22a..50d1e34 100644
+--- a/src/wl/sys/wl_cfg80211_hybrid.c
++++ b/src/wl/sys/wl_cfg80211_hybrid.c
+@@ -105,20 +105,25 @@ static s32 wl_cfg80211_get_tx_power(struct wiphy *wiphy, struct wireless_dev *wd
+ static s32 wl_cfg80211_get_tx_power(struct wiphy *wiphy, s32 *dbm);
+ #endif
+ 
++#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)
++#define MAYBE_INT_LINK_ID int link_id,
++#else
++#define MAYBE_INT_LINK_ID
++#endif
+ #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 38)
+ static s32 wl_cfg80211_config_default_key(struct wiphy *wiphy,
+-           struct net_device *dev, u8 key_idx, bool unicast, bool multicast);
++           struct net_device *dev, MAYBE_INT_LINK_ID u8 key_idx, bool unicast, bool multicast);
+ #else
+ static s32 wl_cfg80211_config_default_key(struct wiphy *wiphy,
+            struct net_device *dev, u8 key_idx);
+ #endif
+ #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 37)
+ static s32 wl_cfg80211_add_key(struct wiphy *wiphy, struct net_device *dev,
+-           u8 key_idx, bool pairwise, const u8 *mac_addr, struct key_params *params);
++           MAYBE_INT_LINK_ID u8 key_idx, bool pairwise, const u8 *mac_addr, struct key_params *params);
+ static s32 wl_cfg80211_del_key(struct wiphy *wiphy, struct net_device *dev,
+-           u8 key_idx, bool pairwise, const u8 *mac_addr);
++           MAYBE_INT_LINK_ID u8 key_idx, bool pairwise, const u8 *mac_addr);
+ static s32 wl_cfg80211_get_key(struct wiphy *wiphy, struct net_device *dev,
+-           u8 key_idx, bool pairwise, const u8 *mac_addr,
++           MAYBE_INT_LINK_ID u8 key_idx, bool pairwise, const u8 *mac_addr,
+            void *cookie, void (*callback) (void *cookie, struct key_params *params));
+ #else
+ static s32 wl_cfg80211_add_key(struct wiphy *wiphy, struct net_device *dev,
+@@ -1165,7 +1170,7 @@ static s32 wl_cfg80211_get_tx_power(struct wiphy *wiphy, s32 *dbm)
+ #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 38)
+ static s32
+ wl_cfg80211_config_default_key(struct wiphy *wiphy,
+-	struct net_device *dev, u8 key_idx, bool unicast, bool multicast)
++	struct net_device *dev, MAYBE_INT_LINK_ID u8 key_idx, bool unicast, bool multicast)
+ #else
+ static s32
+ wl_cfg80211_config_default_key(struct wiphy *wiphy,
+@@ -1190,7 +1195,7 @@ wl_cfg80211_config_default_key(struct wiphy *wiphy,
+ #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 37)
+ static s32
+ wl_cfg80211_add_key(struct wiphy *wiphy, struct net_device *dev,
+-                    u8 key_idx, bool pairwise, const u8 *mac_addr, struct key_params *params)
++                    MAYBE_INT_LINK_ID u8 key_idx, bool pairwise, const u8 *mac_addr, struct key_params *params)
+ #else
+ static s32
+ wl_cfg80211_add_key(struct wiphy *wiphy, struct net_device *dev,
+@@ -1311,7 +1316,7 @@ wl_cfg80211_add_key(struct wiphy *wiphy, struct net_device *dev,
+ #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 37)
+ static s32
+ wl_cfg80211_del_key(struct wiphy *wiphy, struct net_device *dev,
+-                    u8 key_idx, bool pairwise, const u8 *mac_addr)
++                    MAYBE_INT_LINK_ID u8 key_idx, bool pairwise, const u8 *mac_addr)
+ #else
+ static s32
+ wl_cfg80211_del_key(struct wiphy *wiphy, struct net_device *dev,
+@@ -1354,7 +1359,7 @@ wl_cfg80211_del_key(struct wiphy *wiphy, struct net_device *dev,
+ #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 37)
+ static s32
+ wl_cfg80211_get_key(struct wiphy *wiphy, struct net_device *dev,
+-                    u8 key_idx, bool pairwise, const u8 *mac_addr, void *cookie,
++                    MAYBE_INT_LINK_ID u8 key_idx, bool pairwise, const u8 *mac_addr, void *cookie,
+                     void (*callback) (void *cookie, struct key_params * params))
+ #else
+ static s32
+-- 
+2.37.2
+

